be/src/util/zip_util.cpp

```diff
@@ -0,0 +1,154 @@
+// under the License.
+
+
+#include "util/zip_util.h"
+#include "util/file_utils.h"
+#include "gutil/strings/substitute.h"
+#include "gutil/strings/util.h"
+#include "env/env.h"
+#include "util/time.h"
+
+namespace doris {
+
+#define DEFAULT_FILE_NAME_SIZE 256
+#define DEFAULT_UNZIP_BUFFER 1048576 // 1MB
+
+#define BREAK_IF_STATUS_ERROR(stmt) { \
+        st = (stmt);                  \
+        if (UNLIKELY(!st.ok())) {     \
+            break;                    \
+        }                             \
+}
+
+using namespace strings;
+
+Status ZipFile::close() {
+    if (_zip_file == nullptr) {
+        unzClose(_zip_file);
+    }
+
+    return Status::OK();
+}
+
+Status ZipFile::extract(const std::string& target_path, const std::string& dir_name) {
+    // check zip file
+    _zip_file = unzOpen64(_zip_path.c_str());
+    if (_zip_file == nullptr) {
+        return Status::InvalidArgument("open zip file: " + _zip_path + " error");
+    }
+
+    unz_global_info64 global_info;
+    int err = unzGetGlobalInfo64(_zip_file, &global_info);
+
+    if (err != UNZ_OK) {
+        return Status::IOError(strings::Substitute("read zip file info $0 error, code: $1", _zip_path, err));
+    }
+
+    // 0.check target path
+    std::string target = target_path + "/" + dir_name;
+    if (FileUtils::check_exist(target)) {
+        return Status::AlreadyExist("path already exists: " + target);
+    }
+
+    // 1.create temp directory
+    std::string temp = target_path + "/.tmp_" + std::to_string(GetCurrentTimeMicros()) + "_" + dir_name;
+    RETURN_IF_ERROR(FileUtils::create_dir(temp));
+
+    // 2.unzip to temp directory
+    Status st;
+    for (int i = 0; i < global_info.number_entry; ++i) {
+        st = extract_file(_zip_file, temp);
+        if (!st.ok()) {
+            FileUtils::remove_all(temp);
+            return st;
+        }
+        unzGoToNextFile(_zip_file);
+    }
+
+    // 3.move to target directory
+    st = Env::Default()->rename_file(temp, target);
+    if (!st.ok()) {
+        FileUtils::remove_all(temp);
+        return st;
+    }
+    return Status::OK();
+
+}
+
+Status ZipFile::extract_file(unzFile zfile, const std::string& target_path) {
+    int err;
+    char file_name[DEFAULT_FILE_NAME_SIZE];
+
+    unz_file_info64 file_info_inzip;
+
+    err = unzGetCurrentFileInfo64(zfile, &file_info_inzip, file_name, DEFAULT_FILE_NAME_SIZE,
+                                  nullptr, 0, nullptr, 0);
+
+    if (err != UNZ_OK) {
+        return Status::IOError(strings::Substitute("read zip file info error, code: $0", err));
+    }
+
+    // is directory, mkdir
+    std::string path = target_path + "/" + std::string(file_name);
+
+    if (HasSuffixString(file_name, "/") || HasSuffixString(file_name, "\\")) {
+        FileUtils::create_dir(path);
+
+        return Status::OK();
+    }
+
+    // is file, unzip
+    Status st = Status::OK();
+    do {
+        err = unzOpenCurrentFile(zfile);
```

Comment:
we can use a bool flag to indicate whether to close it when destructing.

Reply:
I think it's not a good idea, because the variable(zfile) will be change in the outer-loop(unzGoToNextFile....) and the method(unzReadCurrentFile) will update zfile's data. It's may be error if replay the loop to close the variable when destructing.